src/mongo/db/write_concern_options.h

```diff
@@ -127,52 +121,92 @@ struct WriteConcernOptions {
      *      - Implicit default majority WC is being used.
      */
     bool isImplicitDefaultWriteConcern() const {
-        return usedDefaultConstructedWC || _provenance.isImplicitDefault();
+        return isDefaultConstructed() || _provenance.isImplicitDefault();
     }
 
-    SyncMode syncMode;
+    /**
+     * True if the default constructed WC ({w:1}) was used.
+     *      - Implicit default WC when value of w is {w:1}.
+     *      - Internal commands set empty WC ({writeConcern: {}}), then default constructed WC (w:1)
+     *        is used.
+     * False otherwise:
+     *      - Implicit default WC when value of w is {w:"majority"}.
+     *      - Cluster-wide WC.
+     *          - with (w) value set, for example ({writeConcern: {w:1}}).
+     *          - without (w) value set, for example ({writeConcern: {j: true}}).
+     *      - Client-supplied WC.
+     *          - with (w) value set, for example ({writeConcern: {w:1}}).
+     *          - without (w) value set, for example ({writeConcern: {j: true}}).
+     */
+    bool isDefaultConstructed() const {
+        return _defaultConstructed;
+    }
+
+    /**
+     * Used only for tracking opWriteConcernCounters metric.
+     * True if the (w) value of the write concern used is not set explicitly by client:
+     *      - Default constructed WC ({w:1})
+     *      - Implicit default majority WC.
+     *      - Cluster-wide WC.
+     *          - with (w) value set, for example ({writeConcern: {w:1}}).
+     *          - without (w) value set, for example ({writeConcern: {j: true}}).
+     *      - Client-supplied WC without (w) value set, for example ({writeConcern: {j: true}}).
+     *      - Internal commands set empty WC ({writeConcern: {}}).
+     */
+    bool hasExplicitWValue() const {
+        return _explicitWValue;
+    };
+
+    SyncMode syncMode() const {
+        return _syncMode;
+    }
 
     // The w parameter for this write concern. The wMode represents the string format and
     // takes precedence over the numeric format wNumNodes.
-    int wNumNodes;
-    std::string wMode;
+    int wNumNodes() const {
+        return _wNumNodes;
+    }
+
+    StringData wMode() const {
+        return _wMode;
+    }
+
+    int wTimeout() const {
+        return _wTimeout;
+    }
+
+    Date_t wDeadline() const {
+        return _wDeadline;
+    }
+
+    // TODO: this is only used in one test
+    // `NodeReturnsWriteConcernFailedWhenAWriteConcernTimesOutBeforeBeingSatisified`
+    void setDeadline_forTest(const Date_t& deadline) {
+        _wDeadline = deadline;
+    }
 
-    // Timeout in milliseconds.
-    int wTimeout;
+    CheckCondition checkCondition() const {
+        return _checkCondition;
+    }
+
+private:
+    friend class repl::ReplicationCoordinatorImpl;
```

Comment:
The `ReplicationCoordinator` does serious surgery to `WriteConcernOptions`
  - [populating unset syncModes](https://github.com/mongodb/mongo/blob/b7181e6813fbc771f4225d8161b4bd7302630c5c/src/mongo/db/repl/replication_coordinator_impl.cpp#L5892-L5904)
  - [setting a checkCondition](https://github.com/mongodb/mongo/blob/b7181e6813fbc771f4225d8161b4bd7302630c5c/src/mongo/db/repl/replication_coordinator_impl.cpp#L3234-L3240)

These aren't changes that are easy to support through API (should we add `checkCondition` to  a constructor, or support it in IDL?), so I made the decision to only allow the `ReplicationCoordinator` to mess with the internals of a write concern. I think we should also reconsider this, but in some future ticket.

Reply:
Billy thinks that we should have deprecated mutators that the ReplicationCoordinator* classes can use (and document them as deprecated in comments) instead of giving them free reign with a friend declaration.